### PR TITLE
[core] asset publishing

### PR DESCRIPTION
**Pull request by kouks**
So I came up with this rather simple proposal on how to handle asset publishing. I like that the core package has no knowledge of `VueRotuer` so it could technically be used with any frontend framework as long as it defines routes via objects. This could also be changed for callbacks in future versions for more flexibility.

You would define your published routes in the provider `register` method using a reference to `this.router` which is a very simple class that only holds the global router.

```typescript
  register () : void {
    this.router.add([
      { path: '/', name: 'home', component: Home },
      ...
    ])
  }
```

Then in the exteranto skeleton app, the vue `router.js` would define routes like this:

```typescript
export default new VueRouter({
  routes: Router.add([
    {
      path: '/about',
      component: About
    }
  ]).get()
})
```

Referencing our `Router` from `@exteranto/core`.

Any ideas?

### CODE REVIEW

**Comment by kouks** [comment]
There could also be best practices like
- each package should define a child route tree with a unique path (like `/package-name/route1`)
- each route should be referenced by `name` rather than path

**Comment by bausano** [comment]
I agree that we should leave the assets binding as simple as possible (like the key value Router service). If a project has a standardized way of managing layouts and they can be referenced just in templates (in Vue it is possible, so I think is in Angular), this will do for the exteranto part.

I am just curious if it wouldn't be more in a spirit of Exteranto to make the Router a singleton annotated service bound in the IoC container rather than a class with static properties.

Often in apps you want to include certain components within other views based on context (e.g. profile settigs for each module in a single path, ...). That could be solved with a layout that let's you for example insert array of compenents before or after the main view. However, in order to register these `pluggable` compenents in service providers, you need a functionality to `edit` path. Let's take following example:

```
// Some core module registers this path:
{
  path: '/settings',
  component: Settigs
}
``` 

```
// Another module wants to alter the path so that
// the front end layout service could insert the pluggable:
{
  path: '/settigs',
  component: Settigs,
  pluggable: [
    {
      where: 'before',
      priority: 3,
      component: DropListSettings
    }
  ]
}
```

This is an abstract example and there might be different and better use cases. The guts of it is that I suggest to add an `edit` method.
```
/**
 * @param {string} name Route to edit.
 * @param {function} fn Takes one argument, the current state of a value of key "name".
 *                                     That value is rewritten with whatever this function returns.
 */
edit (name: string, fn: (current: any) -> any) : void
```
Then we can do 
````
router.edit('settings', (route) => {
  // This is just an implementation detail so it looks ugly but yeah, still can do a Route class or something.
  if (!Array.isArray(path.pluggable)) {
    route.pluggable = []
  }

  route.pluggable.push(DropListComponent)

  // Or we can introduce side effects and get rid of this line.
  return path
}
```

File: lib/core/src/Router.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * Add routes to the global collection.
+   *
+   * @param {any[]} routes
+   * @param {Script} script
+   */
+  public static add (routes: any[], script?: Script) : typeof Router {
+    if (script === undefined || script === this.script) {
+      this.routes = this.routes.concat(routes)
```
Comment (bausano) [inline lib/core/src/Router.ts]: If we're going to do the `edit` functionality, I reckon
`this.routes.push(...routes)`
is preferable?

**Comment by kouks** [comment]
I though about this and:
- binding the `Router` to the container seems weird to me as in the actual Vue router in the skeleton application I would have to import `Container` and resolve the binding. It would also likely be the only place this is done. I don't want users to `@Autowire` the `Router` as it has no effect editing it after the application is booted.
- editing routes might be tricky as the order of registering providers would matter. What if the user registers the provider that edits a route before registering the provider that initially creates the route? (order in the config file).

**Comment by bausano** [comment]
Agreed, let's keep `Router` out of IoC.

As for the routes editting, I don't think having the providers list ordered would be an issue, but it doesn't even have to be that way. If the edits would be anonymous functions, we could just store them into an array, wait for all packages to call their `set` functions on Router in booting process and then execute all `edit` functions. This way it wouldn't even matter if a route was registered at all (you would just skip it).

I'm not sure what do you mean by exporting the views for users to edit them. However I don't see this as Exteranto issue, this would be something the extension developers would decide themselves for their use case. By giving them an easy way to edit routes, they have more freedom.

---
For our internal use case, I wouldn't call it dependency on one another, but rather extensibility of each other:

Droplist doesn't really care if there is a settings page. Only if there is it edits that route by plugging its component into it, otherwise that edit function is skipped. The settings route doesn't depend on droplist module either. Also, we are building that extension in a closed, tightly connected environment, which decreases the requirements on how Lego-like our modules architecture actually has to be.

Nevertheless, I'd leave Exteranto with a `Router` that has just `get`, `set` and `edit`. The rest imo should depend on specific extension you're building.

**Comment by bausano** [comment]
Also by executing edit functions last, we can express following theoretical use case:

Droplist extends deallist and settings.
Droplist adds its own route.
Droplist extends menu by plugging its icon there.
Notifications extend droplist.

If you removed Droplist plugging from the build, it would have desired effect - droplist wouldn't be included in the build and everything would work as expected.

Maybe we could call the function `extend` instead of `edit`.

**Comment by bausano** [comment]
Another way to phrase it would be:

Edit functions define relationships among modules.

**Comment by bausano** [comment]
closes #14

**Pull request merged**